Logging/LogHandler/FileLogHandler.cs

```diff
@@ -165,52 +176,107 @@ protected override void DisposeSelf()
                 callerDerivedTypeName, filename, callerName, callerLine
             );
 
-            m_Writer.WriteLine($"{timestamp}{logLevel}{context}{message}");
+            string log = $"{timestamp}{logLevel}{context}{message}";
 
-            if (m_ShouldRotate && m_Writer.BaseStream.Length > RotateFileSizeLimit)
+            if (m_RotateFileSizeLimit != null)
             {
-                this.RotateFiles();
+                int logLength = Encoding.Unicode.GetByteCount(log);
+                long fileLength = m_Writer.BaseStream.Length;
+
+                if ((fileLength + logLength) > m_RotateFileSizeLimit)
+                {
+                    if (logLength > m_RotateFileSizeLimit)
+                    {
+                        throw new Exception("Message size exceeds file size limit, logging failed " +
```

Comment:
Apologies if I asked for an exception here. We should probably just log an error or cull the message and append some `{culledMessage}...(message longer than file size limit)` type message.

No sense in crashing the application because a log that's too long slipped in. Thinking of cases where we log serialized JSON state and it's fine until suddenly it's too big some day and isn't ok.

Same goes for crazy stacks.

Reply:
I meant to wrap all the exceptions in `#if DEBUG` but yeah, I can have it trim the log and append a message about it being truncated. Hard to log a warning though, as that would just be pushed to the next log file, and may start an infinite truncation warning loop if you set a really small file size limit by accident.

Reply:
The problem with truncating a message to a certain size in bytes is that char count != byte count. Maybe you have lots of 2-byte unicode characters in your log. I can think of two ways to handle this:

1. Convert to byte array, truncate, convert to string
    - this risks cutting a unicode char in half and "corrupting" the last visible character in the log.
2. Truncate the string to `targetByteCount` chars to start, then continue trimming one char at a time and checking `Encoding.Unicode.GetByteCount(log)` until it's under the limit
    - gross but functional, and very very rarely used, unless you've set your file size limit to 10KB and are logging massive JSON files or something

Reply:
For option 2, would it be better to
 - Go backward one character at a time
 - Measure the byte size of that one character
 - Repeat until your threshold is met?

**Option 3** 
On the other hand, accuracy isn't the most important thing in the world and most characters are 1-byte in UTF-8 (the ASCII range) so you could assume that 1 character == 1 byte and worst case you'd remove 4x the number of characters you needed to. 🤷 

I'm goo with either Option 2 or 3. Just throw a comment in there describing what we're doing.

Life would be easy if we just ASCII encoded but that could be restrictive...and would support emojis.

Reply:
We definitely need unicode support for emojis... and non-English names I guess. 😉

Things got hairy here, but I've pushed a fairly reasonable solution:
- Each log goes through `m_Writer.Encoding.GetMaxByteCount(...)` to determine if we're anywhere near the file size limit
  - _This is a cheap check, basically assuming every char is the maximum byte count for the current system's encoding_
- If the next log may go over the limit, call the more expensive `m_Writer.Encoding.GetByteCount()`
- Remove at least enough chars to guarantee this log will be under the limit (including appended truncation message)
  - If the culled chars were all ASCII, the file will be exactly the file size limit
  - If any culled chars were two-byte unicode, we're just trimming a bit extra and coming in a little under the limit

Tested by creating a `FileLogHandler` with a size limit of 1024 bytes, and logging a 1000-char string with some unicode chars in it. With the addition of the log context and timestamp (and a system-specific newline added by the writer when you call `WriteLine(...)`) this goes well over the limit, is truncated, and has a truncation message appended, resulting in a file exactly 1KB (1024 B) in size on my PC.